receiver/oktareceiver/logs.go

```diff
@@ -0,0 +1,236 @@
+// Copyright observIQ, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package oktareceiver
+
+import (
+	"context"
+	"encoding/json"
+	"io"
+	"net/http"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/okta/okta-sdk-golang/v2/okta"
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/consumer"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/plog"
+	"go.uber.org/zap"
+)
+
+// oktaMaxLimit maximum number of log objects returned in one call to Okta API
+var oktaMaxLimit = 1000
+
+type oktaLogsReceiver struct {
+	cfg      Config
+	client   httpClient
+	consumer consumer.Logs
+	logger   *zap.Logger
+	nextURL  string
+	cancel   context.CancelFunc
+	wg       *sync.WaitGroup
+}
+
+type httpClient interface {
+	Do(req *http.Request) (*http.Response, error)
+	CloseIdleConnections()
+}
+
+// newOktaLogsReceiver returns a newly configured oktaLogsReceiver
+func newOktaLogsReceiver(cfg *Config, logger *zap.Logger, consumer consumer.Logs) (*oktaLogsReceiver, error) {
+	return &oktaLogsReceiver{
+		cfg:      *cfg,
+		client:   http.DefaultClient,
+		consumer: consumer,
+		logger:   logger,
+		wg:       &sync.WaitGroup{},
+	}, nil
+}
+
+func (r *oktaLogsReceiver) Start(_ context.Context, _ component.Host) error {
+	ctx, cancel := context.WithCancel(context.Background())
+	r.cancel = cancel
+	r.wg.Add(1)
+	go r.startPolling(ctx)
+	return nil
+}
+
+func (r *oktaLogsReceiver) startPolling(ctx context.Context) {
+	defer r.wg.Done()
+	t := time.NewTicker(r.cfg.PollInterval)
+
+	err := r.poll(ctx)
+	if err != nil {
+		r.logger.Error("there was an error during the first poll", zap.Error(err))
+	}
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-t.C:
+			err := r.poll(ctx)
+			if err != nil {
+				r.logger.Error("there was an error during the poll", zap.Error(err))
+			}
+		}
+	}
+}
+
+func (r *oktaLogsReceiver) poll(ctx context.Context) error {
+	logEvents := r.getLogs(ctx)
+	observedTime := pcommon.NewTimestampFromTime(time.Now())
+	logs := r.processLogEvents(observedTime, logEvents)
+	if logs.LogRecordCount() > 0 {
+		if err := r.consumer.ConsumeLogs(ctx, logs); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *oktaLogsReceiver) getLogs(ctx context.Context) []*okta.LogEvent {
+	var logs []*okta.LogEvent
```

Comment:
Any reason for *okta.LogEvent instead of just okta.LogEvent?

Reply:
From my understanding it would be less data to pass around between functions. Maybe my pointer knowledge is off

Reply:
I wouldn't worry about copying between functions here, slices are already pointers so passing around a `[]*okta.LogEvent` has the same overhead as passing around a `[]okta.LogEvent`.